### PR TITLE
Remove xcm on_runtime_upgrade pallet hook

### PR DESCRIPTION
**Pull request by pgherveou**
Move the pallet `on_runtime_upgrade` hook logic into migration.rs

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2839855

File: xcm/pallet-xcm/src/migration.rs
Comment (bkchr) [inline xcm/pallet-xcm/src/migration.rs:59]:
```suggestion
				StorageVersion::new(1).put::<Pallet<T>>();
```

**Review by bkchr — APPROVED** [review]

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2851130

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2914345

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2944715

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2944951

File: xcm/pallet-xcm/src/lib.rs
Comment (pgherveou) [reply xcm/pallet-xcm/src/lib.rs:688]: It's hard to check for all of them. Every migration should be tested with try-runtime which will yell at you if there is a storage mismatch for any pallet

File: xcm/pallet-xcm/src/migration.rs
Comment (liamaharon) [inline xcm/pallet-xcm/src/migration.rs:31]: We should remove the `pre_upgrade` hook here, otherwise on-runtime-upgrade will start breaking after this upgrade is enacted on-chain

File: xcm/pallet-xcm/src/migration.rs
Comment (liamaharon) [inline xcm/pallet-xcm/src/migration.rs:40]: Does this make sense to run even if the upgrade is a noop (`StorageVersion::get::<Pallet<T>>() != 0`)?

File: xcm/pallet-xcm/src/lib.rs
Comment (liamaharon) [reply xcm/pallet-xcm/src/lib.rs:688]: Is there a list of runtimes we can check off before merging this?

Comment (pgherveou) [reply xcm/pallet-xcm/src/lib.rs:688]: I initially did that in this PR, but after reviews decided to only apply it to the chain I was upgrading (contracts-rococo)
https://github.com/paritytech/cumulus/pull/2570, (see also other earlier commit where I apply it to all runtimes https://github.com/paritytech/cumulus/pull/2570/commits/d8905708e182ea8b1b094980a806ef0f08b2aa33)

The reasoning is that adding a migration that have already been applied, adds some unnecessary bloat to the wasm runtime, so the best practice should be to test with try-runtime and see what needs to be included.

File: xcm/pallet-xcm/src/migration.rs
Comment (pgherveou) [reply xcm/pallet-xcm/src/migration.rs:40]: I tried not to alter the existing logic here, which I think is fine.
If you run the migration with try-runtime from v1 it will fail because of the assert in the pre-upgrade, indicating that the fix is to remove this migration from the migrations passed to `frame_executive::Executive`

File: xcm/pallet-xcm/src/lib.rs
Comment (pgherveou) [reply xcm/pallet-xcm/src/lib.rs:688]: following up, here since @vstam1  is trying to close all pending issues before the repo migration.
This logic is moved to the Migration Struct that should be included in the `frame_executive::Executive` when a migration needs to be passed. 

My understanding is that `on_runtime_upgrade` hooks are deprecated and the migration logic should move out of the pallet into these Migration structs so that they can be controlled and configured by runtime authors.

This is why I opened this PR in the first place, so that `pallet_xcm` adheres to this new conventions and integrate better with tools like try-runtime

Comment (KiChjang) [reply xcm/pallet-xcm/src/lib.rs:688]: Yes, but what we're saying here in this comment is whether or not the existing runtimes that we care about is already on storage version 1.

Comment (pgherveou) [reply xcm/pallet-xcm/src/lib.rs:688]: I don't think this should impact whether or not we merge this PR? 
The code here was moved to MigrateToV1.
If one runtime is not on v1 then it will run the upgrade by passing MigrateToV1 to frame_executive
if it's already on V1, then you just don't use MigrateToV1.
In both scenarios dry-running the migration with try-runtime will tell you if something is missing or superfluous.

Feel free to close to unblock the repo migration if I got that wrong

File: xcm/pallet-xcm/src/migration.rs
Comment (liamaharon) [reply xcm/pallet-xcm/src/migration.rs:31]: Could you do it now please, or better yet use a `VersionedRuntimeUpgrade` which was designed specifically for these use cases? 

We do not want runtime upgrade hooks to pass or fail depending on whether or not an upgrade has run. The approach of "Remove it next time you run an upgrade" is flakey and will cause our CI to break in the period between the upgrade and removing it.

**Review by liamaharon — CHANGES_REQUESTED** [review]

File: xcm/pallet-xcm/src/migration.rs
Comment (pgherveou) [reply xcm/pallet-xcm/src/migration.rs:31]: so how is that being tested in CI? against which runtime version?

Comment (liamaharon) [reply xcm/pallet-xcm/src/migration.rs:31]: it's tested using `try-runtime-cli` using the current on-chain state. 

see jobs `gitlab-check-runtime-migration-polkadot, `gitlab-check-runtime-migration-westend`, etc

File: xcm/pallet-xcm/src/migration.rs
Comment (liamaharon) [inline xcm/pallet-xcm/src/migration.rs:43]: this will work much better with try-runtime-cli. thanks. 👌

**Review by liamaharon — APPROVED** [review]

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/3333964

**Comment by pgherveou** [comment]
bot merge

**Pull request merged**
